test/jdk/java/nio/channels/FileChannel/TransferFromExtend.java

```diff
@@ -0,0 +1,160 @@
+        }
+        return list.stream();
+    }
+
+    //
+    // transfer size must be greater than the threshold
+    // sun.nio.ch.FileChannelImpl::MAPPED_TRANSFER_THRESHOLD (16K)
+    // for a mapped transfer to be used when direct is unavailable
+    //
+    private static Stream<Arguments> fastParamProvider() {
+        return paramProvider(16*1024 + 1, 500*1024);
+    }
+
+    private static Stream<Arguments> readingByteChannelParamProvider() {
+        return paramProvider(1, 64*1024);
+    }
+
+    //
+    // Test the direct and memory-mapped paths. At present the direct path
+    // is implemented only on Linux. The mapped path will be taken only if
+    // there is no direct path and the size of the transfer is more than
+    // sun.nio.ch.FileChannelImpl::MAPPED_TRANSFER_THRESHOLD (16K).
+    // This method therefore tests the direct path on Linux and the mapped
+    // path on other platforms.
+    //
+    @ParameterizedTest
+    @MethodSource("fastParamProvider")
+    void fromFast(long initialPosition, int bufSize, long offset) throws IOException {
+        Path file = Files.createTempFile(DIR, "foo", "bar");
+        try (FileChannel src = FileChannel.open(file, DELETE_ON_CLOSE,
+                                                READ, WRITE)) {
+            byte[] bytes = new byte[bufSize];
+            RND.nextBytes(bytes);
+            ByteBuffer buf = ByteBuffer.wrap(bytes);
+            int total = 0;
+            while (total < bufSize) {
+                int n = src.write(ByteBuffer.wrap(bytes), 0);
+                assertTrue(n >= 0, n + " < " + 0);
+                total += n;
+            }
+            transferFrom(src, src.size(), initialPosition, offset, bytes);
+        }
+    }
+
+    //
+    // Test the arbitrary source path. This method tests the
+    // generic path on all platforms.
+    //
+    @ParameterizedTest
+    @MethodSource("readingByteChannelParamProvider")
+    void fromReadingByteChannel(long initialPosition, int bufSize, long offset)
+        throws IOException {
+        byte[] bytes = new byte[bufSize];
+        RND.nextBytes(bytes);
+        ByteArrayInputStream bais = new ByteArrayInputStream(bytes);
+        try (ReadableByteChannel src = Channels.newChannel(bais)) {
+            transferFrom(src, bufSize, initialPosition, offset, bytes);
+        }
+    }
+
+    /**
+     * Tests transferring bytes to a FileChannel from a ReadableByteChannel.
+     *
+     * @param src        the source of the bytes to transfer
+     * @param count the  number of bytes to transfer
+     * @param initialPos the position of the target channel before the transfer
+     * @param offset     the offset beyong the end of the target channel
+     * @param bytes      the bytes expected to be transferred
+     */
+    private static void transferFrom(ReadableByteChannel src, long count,
+                                     long initialPos, long offset, byte[] bytes)
+        throws IOException {
```

Comment:
Can you fix the indention here, it's hard to see where the method declaration ends and the body starts.